fennel/test_lib/executor.py

```diff
@@ -458,3 +459,99 @@ def visitFirst(self, obj):
         df = df.sort_values(input_ret.timestamp_field)
         df = df.groupby(obj.keys).first().reset_index()
         return NodeRet(df, input_ret.timestamp_field, input_ret.key_fields)
+
+    def visitWindow(self, obj):
+        input_ret = self.visit(obj.node)
+        if input_ret is None or input_ret.df.shape[0] == 0:
+            return None
+
+        input_df = copy.deepcopy(input_ret.df)
+        input_df = input_df.sort_values(input_ret.timestamp_field)
+
+        def generate_sessions_for_df(
+            df: pd.DataFrame, timestamp_col: str, gap: int, field: str
+        ):
+            # Sort the DataFrame by timestamp_col
+            df = df.sort_values(by=timestamp_col)
+
+            # Initialize lists to store session information
+            session_list: List[dict] = []
+            timestamp_list = []
+            current_session_start_time = None
+            current_session_end_time = None
+            event_count = 0
+
+            # Iterate through the rows of the DataFrame
+            for index, row in df.iterrows():
+                # Check if it's the first event for this user
+                if current_session_start_time is None:
+                    current_session_start_time = row[timestamp_col]
+                    current_session_end_time = row[timestamp_col]
+                    event_count = 1
```

Comment:
I think we could actually create a Session/Window struct that has the apis, add_event and automatically increments event count, and sets start/end timestamp appropriately. We can then use this at line 506 and append it to the session list. 

This struct will also help us later when we want to store the list of events / do tumbling windows etc

Finally when creating the dataframe we can convert it to dictionary using some list comprehension syntax.

Reply:
Added the impl.